packages/core/react/src/WalletProvider.tsx

```diff
@@ -42,41 +43,71 @@ export const WalletProvider: FC<WalletProviderProps> = ({
     onError,
     localStorageKey = 'walletName',
 }) => {
-    const [name, setName] = useLocalStorage<WalletName | null>(localStorageKey, null);
-    const [{ wallet, adapter, ready, publicKey, connected }, setState] = useState(initialState);
-    const [connecting, setConnecting] = useState(false);
-    const [disconnecting, setDisconnecting] = useState(false);
-    const isConnecting = useRef(false);
-    const isDisconnecting = useRef(false);
-    const isUnloading = useRef(false);
+    // When the wallets change, initialize the details
+    const [details, setDetails] = useInitialState(
+        () =>
+            wallets.reduce<Record<WalletName, WalletDetails>>((details, wallet) => {
+                details[wallet.name] = { ready: false };
+                return details;
+            }, {}),
+        [wallets]
+    );
+
+    // When the wallets change, asynchronously update the details
+    useEffect(() => {
+        (async () => {
+            const waiting = wallets;
+            const ready = await Promise.all(wallets.map((wallet) => wallet.adapter.ready()));
+            // If the wallets haven't changed while waiting, update the details
+            if (wallets === waiting) {
+                const details = wallets.reduce<Record<WalletName, WalletDetails>>((details, wallet, index) => {
+                    details[wallet.name] = { ready: ready[index] };
+                    return details;
+                }, {});
+                setDetails(details);
+            }
+        })();
+    }, [wallets]);
```

Comment:
@danmt `details` sucks as a name, but the idea here is to provide an extensible lookup table for wallet metadata. This will be nonbreaking, the UI will still get the full list, and then can sort/filter with whatever custom logic applies. Perhaps the adapter could be expanded on to include information based on device type, etc. that would allow the app to prioritize the wallets.

Reply:
I don't like `details`  as a name either but it does work for this (maybe we could use `metadata` instead)... I like the idea of a lookup table for wallet metadata because it's a non-breaking change, but I do think that having the ready/isMobile/etc... as part of the wallet when loading them is probably easier to reason about. I'm happy with either approach tho but I'm leaning a bit onto adding this information straight to the wallet and avoiding the `details` object...

If we commit to the `details` object, is there a reason why not use a `Map` instead of a plain object?

Reply:
> I do think that having the ready/isMobile/etc... as part of the wallet when loading them is probably easier to reason about

I agree. I haven't come up with a clean way to implement this. Any ideas?

> why not use a Map instead of a plain object?

What's the benefit of using a Map?

Reply:
> I agree. I haven't come up with a clean way to implement this. Any ideas?

What if... the factory functions we have per wallet (i.e. getPhantomWallet) returned a Promise instead, the responsibility of the Promise is to create a new adapter, get the ready status and return a `Wallet` that includes all the values we already have plus `ready` boolean field. This also opens the door for adding more fields in the future (like isMobile).

The reason why I like this idea is that:

1. We call ready once per wallet when loading them
2. For consumers is not a breaking change
3. There's no need for a lookup or maintaining more info in the state.

BTW - I haven't done full-time research on the pros/cons of this strategy, I'd love to hear your angle on this.

> What's the benefit of using a Map?

There's no major benefit, I like them more than objects because it has built-in methods like `get` and `has`, another cool thing is that's iterable and can be called with a `for..of` loop.